test/Microsoft.SqlTools.ServiceLayer.UnitTests/Messaging/MessageDispatcherTests.cs

```diff
@@ -84,5 +86,67 @@ public void OnListenTaskCompletedFaultedTaskTest()
                 Assert.True(handlerCalled);   
             }
         }
+
+        [Test]
+        public void ParallelMessageProcessingTest()
+        {
+            int numOfRequests = 11;
+            int msForEachRequest = 1000;
+            // Without parallel processing, this should take around numOfRequests * msForEachRequest ms to finish.
+            // With parallel process, this should take around 1 * msForEachRequest ms to finish.
+            // The diff should be greater than (numOfRequests - 1) * msForEachRequest ms.
+            Assert.IsTrue(GetTimeToHandleRequests(false, numOfRequests, msForEachRequest) - GetTimeToHandleRequests(true, numOfRequests, msForEachRequest) > msForEachRequest * (numOfRequests - 1));
+        }
+
+
+        /// <summary>
+        /// Gets the time to handle certain amount of requests in ms
+        /// </summary>
+        /// <param name="parallelMessageProcessing">Wheater to enable parallel processing</param>
+        /// <param name="numOfRequests">num of requests to handle</param>
+        /// <param name="msForEachRequest">rough time taken to finish each reqeust in ms</param>
+        /// <returns></returns>
+        private long GetTimeToHandleRequests(bool parallelMessageProcessing, int numOfRequests, int msForEachRequest)
+        {
+            RequestType<int, int> requestType = RequestType<int, int>.Create("test/requestType");
+            var mockChannel = new Mock<ChannelBase>();
+            SemaphoreSlim unfinishedRequestCount = new SemaphoreSlim(numOfRequests);
+            bool okayToEnd = false;
+            mockChannel.Setup(c => c.MessageReader.ReadMessage())
+                .Returns(Task.FromResult(Message.Request("1", "test/requestType", null)));
+            var dispatcher = new MessageDispatcher(mockChannel.Object);
+            dispatcher.ParallelMessageProcessing = parallelMessageProcessing;
+            Stopwatch stopwatch = Stopwatch.StartNew();
+            var handler = async (int _, RequestContext<int> _) =>
+            {
+                Thread.Sleep(msForEachRequest);
+                unfinishedRequestCount.Wait();
```

Comment:
What's the purpose of waiting sync here?
You can await with `WaitAsync()` on it.